src/golang/lib/workflow/engine/aq_engine.go

```diff
@@ -0,0 +1,811 @@
+	storageConfig := workflowDagsToDelete[0].StorageConfig
+	for _, workflowDag := range workflowDagsToDelete {
+		if !reflect.DeepEqual(workflowDag.StorageConfig, storageConfig) {
+			return errors.New("Workflow Dags have mismatching storage config.")
+		}
+	}
+
+	workflow_utils.CleanupStorageFiles(ctx, &storageConfig, storagePaths)
+
+	// Delete the cron job if it had one.
+	if workflowObject.Schedule.CronSchedule != "" {
+		cronjobName := shared_utils.AppendPrefix(workflowObject.Id.String())
+		err = eng.cronjobManager.DeleteCronJob(ctx, cronjobName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to delete workflow's cronjob.")
+		}
+	}
+	return nil
+}
+
+func (eng *aqEngine) EditWorkflow(
+	ctx context.Context,
+	workflowId uuid.UUID,
+	workflowName string,
+	workflowDescription string,
+	schedule *workflow.Schedule,
+) error {
+	changes := map[string]interface{}{}
+	if workflowName != "" {
+		changes["name"] = workflowName
+	}
+
+	if workflowDescription != "" {
+		changes["description"] = workflowDescription
+	}
+
+	if schedule.Trigger != "" {
+		cronjobName := shared_utils.AppendPrefix(workflowId.String())
+		err := eng.updateWorkflowSchedule(ctx, workflowId, cronjobName, schedule)
+		if err != nil {
+			return errors.Wrap(err, "Unable to update workflow schedule.")
+		}
+		changes["schedule"] = schedule
+	}
+
+	_, err := eng.WorkflowWriter.UpdateWorkflow(ctx, workflowId, changes, eng.database)
+	if err != nil {
+		return errors.Wrap(err, "Unable to update workflow.")
+	}
+
+	return nil
+}
+
+func (eng *aqEngine) cleanupWorkflow(ctx context.Context, workflowDag dag_utils.WorkflowDag) {
+	for _, op := range workflowDag.Operators() {
+		op.Finish(ctx)
+	}
+}
+
+func (eng *aqEngine) execute(
+	ctx context.Context,
+	workflowDag dag_utils.WorkflowDag,
+	workflowRunMetadata *workflowRunMetadata,
+	timeConfig *AqueductTimeConfig,
+	jobManager job.JobManager,
+	shouldPersistResults bool,
+) error {
+	// These are the operators of immediate interest. They either need to be scheduled or polled on.
+	inProgressOps := workflowRunMetadata.inProgressOps
+	completedOps := workflowRunMetadata.completedOps
+	dag := workflowDag
+	opToDependencyCount := workflowRunMetadata.opToDependencyCount
+
+	// Kick off execution by starting all operators that don't have any inputs.
+	for _, op := range dag.Operators() {
+		if opToDependencyCount[op.ID()] == 0 {
+			inProgressOps[op.ID()] = op
+		}
+	}
+
+	if len(inProgressOps) == 0 {
+		return errors.Newf("No initial operators to schedule.")
+	}
+
+	// Wait a little bit for all active operators to finish before exiting on failure.
+	defer waitForInProgressOperators(ctx, inProgressOps, timeConfig.OperatorPollInterval, timeConfig.CleanupTimeout)
+
+	start := time.Now()
+
+	for len(inProgressOps) > 0 {
+		if time.Since(start) > timeConfig.ExecTimeout {
+			return errors.New("Reached timeout waiting for workflow to complete.")
+		}
+
+		for _, op := range inProgressOps {
+			execState, err := op.GetExecState(ctx)
```

Comment:
For `op.GetExecState()` and and `op.Launch()`, which doesn't exist but is introduced in https://github.com/aqueducthq/aqueduct/pull/296, should we pass the job manager into them as arguments, or continue to keep the job manager in the op constructors.

Reply:
We can keep them in the op constructors.